pallets/rewards/src/mechanism/gap.rs

```diff
@@ -0,0 +1,377 @@
+use cfg_traits::ops::ensure::{
+	EnsureAdd, EnsureAddAssign, EnsureFixedPointNumber, EnsureFrom, EnsureInto, EnsureSub,
+	EnsureSubAssign,
+};
+use frame_support::{pallet_prelude::*, traits::tokens};
+use num_traits::Signed;
+pub use pallet::*;
+use sp_runtime::{
+	traits::{One, Saturating, Zero},
+	ArithmeticError, FixedPointNumber, FixedPointOperand,
+};
+
+use super::{MoveCurrencyError, RewardMechanism};
+
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+#[cfg_attr(test, derive(PartialEq, Clone))]
+pub struct Group<T: Config> {
+	total_stake: T::Balance,
+	pending_total_stake: T::Balance,
+	rpt: T::Rate,
+	distribution_id: T::DistributionId,
+}
+
+impl<T: Config> Default for Group<T> {
+	fn default() -> Self {
+		Self {
+			total_stake: T::Balance::zero(),
+			pending_total_stake: T::Balance::zero(),
+			rpt: T::Rate::zero(),
+			distribution_id: T::DistributionId::default(),
+		}
+	}
+}
+
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebug)]
+#[cfg_attr(test, derive(PartialEq, Clone))]
+pub struct Account<T: Config> {
+	stake: T::Balance,
+	reward_tally: T::IBalance,
+	pending_stake: T::Balance,
+	distribution_id: T::DistributionId,
+	last_currency_movement: u16,
+}
+
+impl<T: Config> Default for Account<T> {
+	fn default() -> Self {
+		Self {
+			stake: T::Balance::zero(),
+			reward_tally: T::IBalance::zero(),
+			pending_stake: T::Balance::zero(),
+			distribution_id: T::DistributionId::default(),
+			last_currency_movement: 0,
+		}
+	}
+}
+
+impl<T: Config> Account<T> {
+	fn was_distribution(&self, group: &Group<T>, currency: &Currency<T>) -> bool {
+		if self.last_currency_movement as usize == currency.rpt_changes.len() {
+			self.distribution_id != group.distribution_id
+		} else {
```

Comment:
Is the purpose of this else branch to handle when a currency has been moved? Maybe adding some comments around why are we doing what would be useful.

Reply:
With the restriction of only moving after distributing, I think I can refactor this to leave the condition simpler.

Reply:
Done